File: original/src/matrix3d.c
// Copyright (C) 1999 Dav Coleman

// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.

// Contact:   Dav Coleman <dav@danger-island.com>
// Homepage:  http://www.danger-island.com/~dav/g3dav/

#include <math.h>

#include "matrix3d.h"
#include "model3d.h"

// quiver of functions to perform on a g3dav_matrix object.
// oops, i mean structure.

// also a special function to transform a list of vertices
// against a given matrix.

g3dav_matrix* g3dav_matrix_new() {
  g3dav_matrix *m;

  m =  g_malloc(sizeof(g3dav_matrix));
  if (m) {
    unit(m);
  }

  return m;
}

// ********************************************************************
// say no to memory leaks
void g3dav_matrix_destroy(g3dav_matrix *matrix) {
   // destroy the model
   g_free(matrix);
}

/** Scale by f in all dimensions */
void uniform_scale(g3dav_matrix* matrix, gfloat f) {
  matrix->xx *= f;
  matrix->xy *= f;
  matrix->xz *= f;
  matrix->xo *= f;
  matrix->yx *= f;
  matrix->yy *= f;
  matrix->yz *= f;
  matrix->yo *= f;
  matrix->zx *= f;
  matrix->zy *= f;
  matrix->zz *= f;
  matrix->zo *= f;
}

/** Scale along each axis independently */
void independent_scale(g3dav_matrix* matrix, gfloat xf, gfloat yf, gfloat zf) {
  matrix->xx *= xf;
  matrix->xy *= xf;
  matrix->xz *= xf;
  matrix->xo *= xf;
  matrix->yx *= yf;
  matrix->yy *= yf;
  matrix->yz *= yf;
  matrix->yo *= yf;
  matrix->zx *= zf;
  matrix->zy *= zf;
  matrix->zz *= zf;
  matrix->zo *= zf;
}

/** Translate the origin */
void translate(g3dav_matrix* matrix, gfloat x, gfloat y, gfloat z) {
  matrix->xo += x;
  matrix->yo += y;
  matrix->zo += z;
}

/** rotate theta degrees about the y axis */
void yrot(g3dav_matrix* matrix, gdouble theta) {
  gfloat Nxx,Nxy,Nxz,Nxo;
  gfloat Nzx,Nzy,Nzz,Nzo;
  gdouble ct, st;

  theta *= (PI / 180);

  ct = cos(theta);
  st = sin(theta);

  Nxx = (gfloat) (matrix->xx * ct + matrix->zx * st);
  Nxy = (gfloat) (matrix->xy * ct + matrix->zy * st);
  Nxz = (gfloat) (matrix->xz * ct + matrix->zz * st);
  Nxo = (gfloat) (matrix->xo * ct + matrix->zo * st);

  Nzx = (gfloat) (matrix->zx * ct - matrix->xx * st);
  Nzy = (gfloat) (matrix->zy * ct - matrix->xy * st);
  Nzz = (gfloat) (matrix->zz * ct - matrix->xz * st);
  Nzo = (gfloat) (matrix->zo * ct - matrix->xo * st);

  matrix->xo = Nxo;
  matrix->xx = Nxx;
  matrix->xy = Nxy;
  matrix->xz = Nxz;
  matrix->zo = Nzo;
  matrix->zx = Nzx;
  matrix->zy = Nzy;
  matrix->zz = Nzz;
}

/** rotate theta degrees about the x axis */
void xrot(g3dav_matrix* matrix, gdouble theta) {
  gfloat Nyx,Nyy,Nyz,Nyo;
  gfloat Nzx,Nzy,Nzz,Nzo;
  gdouble ct, st;

  theta *= (PI / 180);
  ct = cos(theta);
  st = sin(theta);

  Nyx = (gfloat) (matrix->yx * ct + matrix->zx * st);
  Nyy = (gfloat) (matrix->yy * ct + matrix->zy * st);
  Nyz = (gfloat) (matrix->yz * ct + matrix->zz * st);
  Nyo = (gfloat) (matrix->yo * ct + matrix->zo * st);

  Nzx = (gfloat) (matrix->zx * ct - matrix->yx * st);
  Nzy = (gfloat) (matrix->zy * ct - matrix->yy * st);
  Nzz = (gfloat) (matrix->zz * ct - matrix->yz * st);
  Nzo = (gfloat) (matrix->zo * ct - matrix->yo * st);

  matrix->yo = Nyo;
  matrix->yx = Nyx;
  matrix->yy = Nyy;
  matrix->yz = Nyz;
  matrix->zo = Nzo;
  matrix->zx = Nzx;
  matrix->zy = Nzy;
  matrix->zz = Nzz;
}

/** rotate theta degrees about the z axis */
void zrot(g3dav_matrix* matrix, gdouble theta) {
  gfloat Nxx,Nxy,Nxz,Nxo;
  gfloat Nyx,Nyy,Nyz,Nyo;
  gdouble ct, st;

  theta *= (PI / 180);
  ct = cos(theta);
  st = sin(theta);

  Nyx = (gfloat) (matrix->yx * ct + matrix->xx * st);
  Nyy = (gfloat) (matrix->yy * ct + matrix->xy * st);
  Nyz = (gfloat) (matrix->yz * ct + matrix->xz * st);
  Nyo = (gfloat) (matrix->yo * ct + matrix->xo * st);

  Nxx = (gfloat) (matrix->xx * ct - matrix->yx * st);
  Nxy = (gfloat) (matrix->xy * ct - matrix->yy * st);
  Nxz = (gfloat) (matrix->xz * ct - matrix->yz * st);
  Nxo = (gfloat) (matrix->xo * ct - matrix->yo * st);

  matrix->yo = Nyo;
  matrix->yx = Nyx;
  matrix->yy = Nyy;
  matrix->yz = Nyz;
  matrix->xo = Nxo;
  matrix->xx = Nxx;
  matrix->xy = Nxy;
  matrix->xz = Nxz;
}

// Multiply this matrix by a second: M = M*R 
void multiplyBy(g3dav_matrix* m1, g3dav_matrix* rhs) {
  gfloat lxx = m1->xx * rhs->xx + m1->yx * rhs->xy + m1->zx * rhs->xz;
  gfloat lxy = m1->xy * rhs->xx + m1->yy * rhs->xy + m1->zy * rhs->xz;
  gfloat lxz = m1->xz * rhs->xx + m1->yz * rhs->xy + m1->zz * rhs->xz;
  gfloat lxo = m1->xo * rhs->xx + m1->yo * rhs->xy + m1->zo * rhs->xz + rhs->xo;

  gfloat lyx = m1->xx * rhs->yx + m1->yx * rhs->yy + m1->zx * rhs->yz;
  gfloat lyy = m1->xy * rhs->yx + m1->yy * rhs->yy + m1->zy * rhs->yz;
  gfloat lyz = m1->xz * rhs->yx + m1->yz * rhs->yy + m1->zz * rhs->yz;
  gfloat lyo = m1->xo * rhs->yx + m1->yo * rhs->yy + m1->zo * rhs->yz + rhs->yo;

  gfloat lzx = m1->xx * rhs->zx + m1->yx * rhs->zy + m1->zx * rhs->zz;
  gfloat lzy = m1->xy * rhs->zx + m1->yy * rhs->zy + m1->zy * rhs->zz;
  gfloat lzz = m1->xz * rhs->zx + m1->yz * rhs->zy + m1->zz * rhs->zz;
  gfloat lzo = m1->xo * rhs->zx + m1->yo * rhs->zy + m1->zo * rhs->zz + rhs->zo;

  m1->xx = lxx;
  m1->xy = lxy;
  m1->xz = lxz;
  m1->xo = lxo;

  m1->yx = lyx;
  m1->yy = lyy;
  m1->yz = lyz;
  m1->yo = lyo;

  m1->zx = lzx;
  m1->zy = lzy;
  m1->zz = lzz;
  m1->zo = lzo;
}

/** Reinitialize to the unit matrix */
void unit(g3dav_matrix* matrix) {
  matrix->xo = 0;
  matrix->xx = 1;
  matrix->xy = 0;
  matrix->xz = 0;
  matrix->yo = 0;
  matrix->yx = 0;
  matrix->yy = 1;
  matrix->yz = 0;
  matrix->zo = 0;
  matrix->zx = 0;
  matrix->zy = 0;
  matrix->zz = 1;
}

// maybe this should go in another source file
/** Transform nvert points from v into tv.  v contains the input
    coordinates in gfloating point.  Three successive entries in
    the array constitute a point.  tv ends up holding the transformed
    points as gdoubles; three successive entries per point */
GSList *transform(g3dav_matrix* matrix, GSList *modelVertices) {
  
  GSList *node;
  three_d_point *old_point;
  three_d_point *new_point;

  gfloat lxx = matrix->xx, lxy = matrix->xy, lxz = matrix->xz, lxo = matrix->xo;
  gfloat lyx = matrix->yx, lyy = matrix->yy, lyz = matrix->yz, lyo = matrix->yo;
  gfloat lzx = matrix->zx, lzy = matrix->zy, lzz = matrix->zz, lzo = matrix->zo;

  GSList *tmodel = NULL;

  for (node = modelVertices; node; node = node->next) {
    old_point = node->data;
    new_point = g_malloc(sizeof(three_d_point));
    if (new_point) {
      new_point->x = (gdouble) (old_point->x * lxx + 
                                old_point->y * lxy + 
                                old_point->z * lxz + lxo);

      new_point->y = (gdouble) (old_point->x * lyx + 
                                old_point->y * lyy + 
                                old_point->z * lyz + lyo);

      new_point->z = (gdouble) (old_point->x * lzx + 
                                old_point->y * lzy + 
                                old_point->z * lzz + lzo);

      //g_print( "adding (%0.2f,%0.2f,%0.2f)\n", point->x, point->y, point->z );

      tmodel = g_slist_append( tmodel, new_point );
    } else {
      g_warning( "unable to allocate memory for new point (%0.2f,%0.2f,%0.2f)\n", 
                 new_point->x, new_point->y, new_point->z );
    }          
  }
  return tmodel;
}

File: macos/Sources/G3DAVApp/Matrix3D.swift
import Foundation

struct Matrix3D {
    var xx: Double = 1
    var xy: Double = 0
    var xz: Double = 0
    var xo: Double = 0

    var yx: Double = 0
    var yy: Double = 1
    var yz: Double = 0
    var yo: Double = 0

    var zx: Double = 0
    var zy: Double = 0
    var zz: Double = 1
    var zo: Double = 0

    static var identity: Matrix3D { Matrix3D() }

    mutating func reset() {
        self = Matrix3D.identity
    }

    mutating func uniformScale(_ factor: Double) {
        xx *= factor
        xy *= factor
        xz *= factor
        xo *= factor
        yx *= factor
        yy *= factor
        yz *= factor
        yo *= factor
        zx *= factor
        zy *= factor
        zz *= factor
        zo *= factor
    }

    mutating func translate(x: Double, y: Double, z: Double) {
        xo += x
        yo += y
        zo += z
    }

    mutating func rotateY(_ thetaDegrees: Double) {
        let theta = thetaDegrees * Double.pi / 180
        let ct = cos(theta)
        let st = sin(theta)

        let nxx = xx * ct + zx * st
        let nxy = xy * ct + zy * st
        let nxz = xz * ct + zz * st
        let nxo = xo * ct + zo * st

        let nzx = zx * ct - xx * st
        let nzy = zy * ct - xy * st
        let nzz = zz * ct - xz * st
        let nzo = zo * ct - xo * st

        xo = nxo
        xx = nxx
        xy = nxy
        xz = nxz
        zo = nzo
        zx = nzx
        zy = nzy
        zz = nzz
    }

    mutating func rotateX(_ thetaDegrees: Double) {
        let theta = thetaDegrees * Double.pi / 180
        let ct = cos(theta)
        let st = sin(theta)

        let nyx = yx * ct + zx * st
        let nyy = yy * ct + zy * st
        let nyz = yz * ct + zz * st
        let nyo = yo * ct + zo * st

        let nzx = zx * ct - yx * st
        let nzy = zy * ct - yy * st
        let nzz = zz * ct - yz * st
        let nzo = zo * ct - yo * st

        yo = nyo
        yx = nyx
        yy = nyy
        yz = nyz
        zo = nzo
        zx = nzx
        zy = nzy
        zz = nzz
    }

    mutating func rotateZ(_ thetaDegrees: Double) {
        let theta = thetaDegrees * Double.pi / 180
        let ct = cos(theta)
        let st = sin(theta)

        let nyx = yx * ct + xx * st
        let nyy = yy * ct + xy * st
        let nyz = yz * ct + xz * st
        let nyo = yo * ct + xo * st

        let nxx = xx * ct - yx * st
        let nxy = xy * ct - yy * st
        let nxz = xz * ct - yz * st
        let nxo = xo * ct - yo * st

        yo = nyo
        yx = nyx
        yy = nyy
        yz = nyz
        xo = nxo
        xx = nxx
        xy = nxy
        xz = nxz
    }

    mutating func multiply(by rhs: Matrix3D) {
        let lxx = xx * rhs.xx + yx * rhs.xy + zx * rhs.xz
        let lxy = xy * rhs.xx + yy * rhs.xy + zy * rhs.xz
        let lxz = xz * rhs.xx + yz * rhs.xy + zz * rhs.xz
        let lxo = xo * rhs.xx + yo * rhs.xy + zo * rhs.xz + rhs.xo

        let lyx = xx * rhs.yx + yx * rhs.yy + zx * rhs.yz
        let lyy = xy * rhs.yx + yy * rhs.yy + zy * rhs.yz
        let lyz = xz * rhs.yx + yz * rhs.yy + zz * rhs.yz
        let lyo = xo * rhs.yx + yo * rhs.yy + zo * rhs.yz + rhs.yo

        let lzx = xx * rhs.zx + yx * rhs.zy + zx * rhs.zz
        let lzy = xy * rhs.zx + yy * rhs.zy + zy * rhs.zz
        let lzz = xz * rhs.zx + yz * rhs.zy + zz * rhs.zz
        let lzo = xo * rhs.zx + yo * rhs.zy + zo * rhs.zz + rhs.zo

        xx = lxx
        xy = lxy
        xz = lxz
        xo = lxo
        yx = lyx
        yy = lyy
        yz = lyz
        yo = lyo
        zx = lzx
        zy = lzy
        zz = lzz
        zo = lzo
    }

    func transform(points: [Point3D]) -> [Point3D] {
        var result: [Point3D] = []
        result.reserveCapacity(points.count)

        let lxx = xx, lxy = xy, lxz = xz, lxo = xo
        let lyx = yx, lyy = yy, lyz = yz, lyo = yo
        let lzx = zx, lzy = zy, lzz = zz, lzo = zo

        for point in points {
            let x = point.x * lxx + point.y * lxy + point.z * lxz + lxo
            let y = point.x * lyx + point.y * lyy + point.z * lyz + lyo
            let z = point.x * lzx + point.y * lzy + point.z * lzz + lzo
            result.append(Point3D(x: x, y: y, z: z))
        }

        return result
    }
}
